### PR TITLE
Interhand3d demo

### PR DESCRIPTION
**Pull request by zengwang430521**
Add demo for interhand 3d task.

### CODE REVIEW

**Comment by jin-s13** [comment]
lint fails.

File: demo/interhand3d_img_demo.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import mmcv
+import numpy as np
+from xtcocotools.coco import COCO
+
+from mmpose.apis import inference_interhand_3d_model, vis_3d_pose_result
+from mmpose.apis.inference import init_pose_model
+from mmpose.core import SimpleCamera
+
+
+def _transfer_interhand_camera_param(interhand_camera_param):
```
Comment (jin-s13) [inline demo/interhand3d_img_demo.py:14]: transfer -> transform

File: demo/interhand3d_img_demo.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        param (dict): camera parameters including:
+            - R: 3x3, camera rotation matrix (camera-to-world)
+            - T: 3x1, camera translation (camera-to-world)
+            - f: 2x1, camera focal length
+            - c: 2x1, camera center
+    """
+    camera_param = {}
+    camera_param['R'] = np.array(interhand_camera_param['camrot']).T
+    camera_param['T'] = np.array(interhand_camera_param['campos'])[:,
+                                                                   np.newaxis]
```
Comment (jin-s13) [inline demo/interhand3d_img_demo.py:35]: np.newaxis -> None

File: mmpose/apis/inference_3d.py [hunk 3]
```diff
@@ -214,6 +215,103 @@ def vis_3d_pose_result(model,
...
+            [255, 144, 30],
+            [153, 51, 0],
+            [153, 51, 0],
+            [153, 51, 0],
+            [153, 51, 0],
+            [255, 51, 13],
+            [255, 51, 13],
+            [255, 51, 13],
+            [255, 51, 13],
+            [103, 37, 8],
```
Comment (ly015) [inline mmpose/apis/inference_3d.py:269]: Consider removing the last comma so multiple elements can be placed in one line.

File: mmpose/apis/inference_3d.py [hunk 4]
```diff
@@ -227,3 +325,161 @@ def vis_3d_pose_result(model,
...
 
     return img
+
+
+def inference_interhand_3d_model(model,
+                                 img_or_path,
+                                 det_results,
+                                 bbox_thr=None,
+                                 format='xywh',
+                                 dataset='TopDownCocoDataset'):
```
Comment (ly015) [inline mmpose/apis/inference_3d.py:335]: data='Interhand3DDataset'

File: demo/interhand3d_img_demo.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            det_result = {
+                'image_name': image_name,
+                'bbox': ann['bbox'],  # bbox format is 'xywh'
+                'camera_para': camera_param,
+                'keypoints_3d_gt': gt_joint
+            }
+            person_results.append(det_result)
+        person_results_list.append(person_results)
+
+    for i, person_results in enumerate(
```
Comment (ly015) [inline demo/interhand3d_img_demo.py:153]: Are these person detection results or hand detection results?

File: demo/interhand3d_img_demo.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            if camera_params is not None:
+                camera_param = {
+                    key: camera_params[capture_key][key][camera_key]
+                    for key in camera_params[capture_key].keys()
+                }
+                camera_param = _transfer_interhand_camera_param(camera_param)
+            else:
+                camera_param = None
+            if gt_joint_params is not None:
+                joint_para = gt_joint_params[capture_key][str(frame_idx)]
```
Comment (ly015) [inline demo/interhand3d_img_demo.py:135]: If "para" means parameter, use "param" or "params"

File: mmpose/core/visualization/image.py [hunk 7]
```diff
@@ -174,14 +177,16 @@ def imshow_keypoints_3d(
 
     for idx, res in enumerate(pose_result):
         kpts = res['keypoints_3d']
+        valid = kpts[:, 3] >= kpt_score_thr
+
```
Comment (ly015) [inline mmpose/core/visualization/image.py:181]: Consider using kpts = kpts[valid] to simplify the code.

File: demo/interhand3d_img_demo.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        help='Json file containing image info.')
+    parser.add_argument(
+        '--camera-param-file',
+        type=str,
+        default=None,
+        help='Camera parameter file for converting 3D pose predictions from '
+        ' the pixel space to camera space. If None, keypoints in pixel space'
+        'will be visualized')
+    parser.add_argument(
+        '--gt-joints-file',
```
Comment (ly015) [inline demo/interhand3d_img_demo.py:61]: Please make sure that the ground truth is optional and the demo script can run without it.

File: demo/interhand3d_img_demo.py (hunk 1 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py:14]: Fixed.

File: demo/interhand3d_img_demo.py (hunk 2 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py:35]: fixed

File: mmpose/apis/inference_3d.py (hunk 3 above)
Comment (zengwang430521) [reply mmpose/apis/inference_3d.py:269]: Fixed

File: mmpose/apis/inference_3d.py (hunk 4 above)
Comment (zengwang430521) [reply mmpose/apis/inference_3d.py]: Fixed

File: demo/interhand3d_img_demo.py (hunk 5 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py]: It's the detection results of hand. I have changed the name to det_results.

File: demo/interhand3d_img_demo.py (hunk 6 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py]: OK, fixed.

File: mmpose/core/visualization/image.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/core/visualization/image.py:181]: If I use kpts = kpts[valid], the limb also need to change.

File: demo/interhand3d_img_demo.py (hunk 8 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py:58]: Yes, this demo can be used without gt_joints and camera parameters.

File: configs/hand/3d_kpt_sview_rgb_img/internet/interhand3d/res50_interhand3d_all_256x256.py [hunk 9]
```diff
@@ -136,7 +136,7 @@
 data_root = 'data/interhand2.6m'
 data = dict(
     samples_per_gpu=16,
-    workers_per_gpu=2,
+    workers_per_gpu=1,
```
Comment (jin-s13) [inline configs/hand/3d_kpt_sview_rgb_img/internet/interhand3d/res50_interhand3d_all_256x256.py:139]: Is this change necessary?

File: mmpose/models/detectors/interhand_3d.py [hunk 10]
```diff
@@ -94,3 +98,116 @@ def forward_test(self, img, img_metas, **kwargs):
...
+                    pose_2d.append(kpts)
+
+            if len(bbox_result) > 0:
+                bboxes = np.vstack(bbox_result)
+                mmcv.imshow_bboxes(
+                    img,
+                    bboxes,
+                    colors='green',
+                    top_k=-1,
+                    thickness=2,
```
Comment (jin-s13) [inline mmpose/models/detectors/interhand_3d.py:181]: Shall we make `colors` and `thickness` args of `show_result`? @ly015

File: mmpose/models/detectors/interhand_3d.py [hunk 11]
```diff
@@ -94,3 +98,116 @@ def forward_test(self, img, img_metas, **kwargs):
...
+                    show=False)
+            if len(pose_2d) > 0:
+                imshow_keypoints(
+                    img,
+                    pose_2d,
+                    skeleton,
+                    kpt_score_thr=0.3,
```
Comment (jin-s13) [inline mmpose/models/detectors/interhand_3d.py:188]: Use `kpt_score_thr` instead of 0.3

File: configs/hand/3d_kpt_sview_rgb_img/internet/interhand3d/res50_interhand3d_all_256x256.py (hunk 9 above)
Comment (zengwang430521) [reply configs/hand/3d_kpt_sview_rgb_img/internet/interhand3d/res50_interhand3d_all_256x256.py:139]: Yes, the data of interhand2.6m v1.0 is too massive. It will cost too much memory if use 2 workers.

File: mmpose/models/detectors/interhand_3d.py (hunk 11 above)
Comment (zengwang430521) [reply mmpose/models/detectors/interhand_3d.py:188]: OK, fixed.

**Review by jin-s13 — APPROVED** [review]

File: demo/docs/3d_hand_demo.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    --json-file tests/data/interhand2.6m/test_interhand2.6m_data.json \
+    --img-root tests/data/interhand2.6m \
+    --camera-param-file tests/data/interhand2.6m/test_interhand2.6m_camera.json \
+    --gt-joints-file tests/data/interhand2.6m/test_interhand2.6m_joint_3d.json \
+    --out-img-root vis_results \
+    --rebase-keypoint-height \
+    --show-ground-truth
+```
+
+```shell
```
Comment (ly015) [inline demo/docs/3d_hand_demo.md:40]: What's the purpose of providing 2 examples? If they are both necessary, maybe add a few descriptions of their differences.

File: demo/interhand3d_img_demo.py (hunk 8 above)
Comment (ly015) [reply demo/interhand3d_img_demo.py:58]: Ok. It's better to add a description to explain that this is optional and what would happen if it is not provided.

File: mmpose/models/detectors/interhand_3d.py (hunk 10 above)
Comment (ly015) [reply mmpose/models/detectors/interhand_3d.py:181]: Ok. Maybe call them "bbox_colors" and "bbox_thickness"?

**Comment by ly015** [comment]
@zengwang430521 Is it possible to further increase the codecov rate?

File: demo/docs/3d_hand_demo.md (hunk 12 above)
Comment (zengwang430521) [reply demo/docs/3d_hand_demo.md:40]: The second example do not use gt keypoints and camera parameters. It's used to show how to use the demo when there is not ground-truth keypoints and camera parameters.

File: demo/interhand3d_img_demo.py (hunk 8 above)
Comment (zengwang430521) [reply demo/interhand3d_img_demo.py:58]: OK, I have added some comments.

File: mmpose/models/detectors/interhand_3d.py (hunk 10 above)
Comment (zengwang430521) [reply mmpose/models/detectors/interhand_3d.py:183]: I have added "bbox_color". And I recommend to remain the name "thickness" because it's also the thickness of 2D limbs.

Comment (ly015) [reply mmpose/models/detectors/interhand_3d.py:183]: In some cases, it's better to have different thicknesses for the bounding box and the skeleton. For example, the skeleton should be thicker to be clearly seen.

File: demo/docs/3d_hand_demo.md (hunk 12 above)
Comment (ly015) [reply demo/docs/3d_hand_demo.md:40]: Ok. What about adding some descriptions like `w/ ground truth`, `w/o ground truth` above each code block?

File: mmpose/models/detectors/interhand_3d.py (hunk 10 above)
Comment (zengwang430521) [reply mmpose/models/detectors/interhand_3d.py:183]: If we want to split thickness of box and skeleton, we also need to modify "show_result()" function of other models. I think we should not add such modification to this PR, which focus on the demo of interhand3d.

File: demo/docs/3d_hand_demo.md (hunk 12 above)
Comment (zengwang430521) [reply demo/docs/3d_hand_demo.md:42]: I have added some descriptions.
